### PR TITLE
JavaScript 测试︰ 单元 vs 功能 vs 集成测试

### PR DESCRIPTION
**Pull request by wild-flame**

### CODE REVIEW

**Comment by marcmoore** [comment]
@sqrthree 认领校对

**Comment by sqrthree** [comment]
@marcmoore 好的

File: TODO/javascript-testing-unit-functional-integration.md [hunk 1]
```diff
@@ -1,116 +1,197 @@
...
+* 译者：[野火wildflame](http://github.com/wild-flame)
 * 校对者：
 
 # JavaScript Testing: Unit vs Functional vs Integration Tests
-
+# JavaScript 测试︰ 单元 vs 功能 vs 集成测试
 
 Unit tests, integration tests, and functional tests are all types of automated tests which form essential cornerstones of continuous delivery, a development methodology that allows you to safely ship changes to production in days or hours rather than months or years.
 
+自动测试包括了单元测试、 集成测试和功能测试。自动测试是一种开发方法，它是持续不断迭代发布产品的基石，使你生产环境里的发布的能够在数小时，数天之内就发生改变，而不是按月或是按年而计。
```
Comment (marcmoore) [inline TODO/javascript-testing-unit-functional-integration.md]: 单元测试、集成测试、功能测试这些自动化测试方法，是项目持续部署的基础。作为一种研发方式，它能帮助你在短时间内安全的发布新特性，不用等上几个月甚至几年。

File: TODO/javascript-testing-unit-functional-integration.md [hunk 2]
```diff
@@ -1,116 +1,197 @@
...
 ### More from this author
 
-
-
 *   [The Future of Programming: WebAssembly & Life After JavaScript](https://www.sitepoint.com/future-programming-webassembly-life-after-javascript/?utm_source=sitepoint&utm_medium=relatedinline&utm_term=&utm_campaign=relatedauthor)
-
-
-
-
+（上面这段文字和本文内容无关，故我打算删掉，请校对过目。）
```
Comment (marcmoore) [inline TODO/javascript-testing-unit-functional-integration.md]: 确实没关系，看了原文，只是一个Block，推荐作者的其他文章

File: TODO/javascript-testing-unit-functional-integration.md [hunk 3]
```diff
@@ -1,116 +1,197 @@
...
 
+你需要了解的第一件事情就是，每一种测试都有它自己的作用。他们都在软件的持续迭代发布中起了重要的作用。
+
 A while back, I was consulting on an ambitious project where the team was having a hard time building a reliable test suite. Because it was hard to use and understand, it rarely got used or maintained.
 
+前阵子，我为一个野心勃勃的项目做了咨询，他们为了建立一个可靠的测试套件可为煞费苦心。因为，由于现有套件很难使用和理解，它基本上没有被维护和使用。
+
 One of the problems I observed with the existing test suite is that it confused unit tests, functional tests, and integration tests. It made absolutely no distinction between any of them.
 
+我观察到的其中一个问题就是，现有套件混淆了单元，功能和集成测试。他们之间没有明显的区别。
```
Comment (marcmoore) [inline TODO/javascript-testing-unit-functional-integration.md]: It made absolutely no distinction between any of them.
让他们变得没有任何区别。

File: TODO/javascript-testing-unit-functional-integration.md [hunk 4]
```diff
@@ -1,116 +1,197 @@
...
+
 Most apps will require both unit tests and functional tests, and many complex apps will also require integration tests.
 
+大多数应用程序都需要单元测试和功能测试，而许多复杂的应用程序在此基础上，还需要集成测试。
+
 *   **Unit tests** ensure that individual components of the app work as expected. Assertions test the component API.
 *   **Integration tests** ensure that component collaborations work as expected. Assertions may test component API, UI, or side-effects (such as database I/O, logging, etc…)
 *   **Functional tests** ensure that the app works as expected from the user’s perspective. Assertions primarily test the user interface.
 
+**单元测试** 用来确保每个组件正常工作 —— 测试组件 API
```
Comment (marcmoore) [inline TODO/javascript-testing-unit-functional-integration.md]: 解释：Test Assertions = Type of testing consisting in verifying if the conditions confirm the product requirements [Wikipedia](https://en.wikipedia.org/wiki/Test_assertion) 
图片

File: TODO/javascript-testing-unit-functional-integration.md [hunk 5]
```diff
@@ -1,116 +1,197 @@
...
+**集成测试** 用来确保不同组件互相合作 —— 测试组件 API, UI, 或者副作用。
+（译注：校对看看 side-effects 这里是啥意思，我知道的 side-effect 是副作用的意思，但我感觉我理解不能。）
```
Comment (marcmoore) [inline TODO/javascript-testing-unit-functional-integration.md]: 因为是集成测试，会测试出一些连带问题，建议翻译成“意外情况”

File: TODO/javascript-testing-unit-functional-integration.md [hunk 6]
```diff
@@ -1,116 +1,197 @@
...
 In order to understand how different tests fit in your software development process, you need to understand that each kind of test has a job to do, and those tests roughly fall into three broad categories:
 
+为了了解如何在您的软件开发过程选择不同测试，你需要了解每种测试它的工作要做，那些测试大致可分为三大类︰
+
 *   User experience tests (end user experience)
 *   Developer API tests (developer experience)
 *   Infrastructure tests (load tests, network integration tests, etc…)
 
+用户体验测试 （最终用户体验）
+* 开发 API 测试 （开发人员体验）
```
Comment (marcmoore) [inline TODO/javascript-testing-unit-functional-integration.md]: 直译的话真的很奇怪
用户体验测试 （针对最终用户）

File: TODO/javascript-testing-unit-functional-integration.md [hunk 7]
```diff
@@ -1,116 +1,197 @@
...
+
 Unit tests are frequently used as a developer feedback mechanism during development. For example, I run lint and unit tests on every file change and monitor the results in a development console which gives me real-time feedback as I’m working.
 
+单元测试常常被用来开发者的反馈机制。比方说，我在工作时，会在我每一次更改之后运行 lint 和单元测试，在 console 里检测运行的结果。
+
 ![Running tests on file change](https://dab1nmslvvntp.cloudfront.net/wp-content/uploads/2016/04/1461566883dev-console-animated-small.gif)
 
 For this to work well, unit tests must run very quickly, which means that asynchronous operations such as network and file I/O should be avoided in unit tests.
 
+为了实现这个目的，单元测试必须很快，也就是说，在单元测试里，一切异步的操作都应该在被避免。
```
Comment (marcmoore) [inline TODO/javascript-testing-unit-functional-integration.md]: 最后一句不通顺。
另外，such as network and file I/O 没有翻译

File: TODO/javascript-testing-unit-functional-integration.md [hunk 8]
```diff
@@ -1,116 +1,197 @@
...
 *   Lightning fast.
 *   A good bug report.
 
+炒鸡简单
+迅雷不及掩耳之势
+有一个好的报告
+
 What do I mean by “a good bug report?”
 
+什么事「好的报告」？
```
Comment (marcmoore) [inline TODO/javascript-testing-unit-functional-integration.md]: 错别字

File: TODO/javascript-testing-unit-functional-integration.md [hunk 9]
```diff
@@ -1,116 +1,197 @@
...
+
 I mean that whatever test runner and assertion library you use, a failing unit test should tell you at a glance:
 
+「好的报告」就是能够一眼就告诉你，
```
Comment (marcmoore) [inline TODO/javascript-testing-unit-functional-integration.md]: 无论你用什么测试软件和assertion库

File: TODO/javascript-testing-unit-functional-integration.md [hunk 10]
```diff
@@ -1,116 +1,197 @@
...
+
 1.  Which component is under test?
 2.  What is the expected behavior?
 3.  What was the actual result?
 4.  What is the expected result?
 5.  How is the behavior reproduced?
 
+1. 什么组件被测试 ing
```
Comment (marcmoore) [inline TODO/javascript-testing-unit-functional-integration.md]: 哪个组件被测试覆盖到

File: TODO/javascript-testing-unit-functional-integration.md [hunk 11]
```diff
@@ -1,116 +1,197 @@
...
+2. 你期望什么行为
+3. 实际是什么结果
+4. 你期望什么结果
+5. 为什么会这样
```
Comment (marcmoore) [inline TODO/javascript-testing-unit-functional-integration.md]: 如何重现

File: TODO/javascript-testing-unit-functional-integration.md [hunk 12]
```diff
@@ -150,23 +231,42 @@ Here are some simple unit test examples from real projects using [Tape](https://
...
 
 1.  The route handler
 2.  The logger
 
+1. 路由处理器（Route handler）
+2. 记录器（Logger）
+
 If we were unit testing the logger, our tests wouldn’t invoke the route handler, or know anything about it.
 
+如果我们对记录器做单元测试。我们不会了解和路由处理相关的任何内容。
```
Comment (marcmoore) [inline TODO/javascript-testing-unit-functional-integration.md]: 如果我们对logger做了单元测试，这些测试是不会调用到route handler，或者说根本就不知道还有个handler。

File: TODO/javascript-testing-unit-functional-integration.md [hunk 13]
```diff
@@ -150,23 +231,42 @@ Here are some simple unit test examples from real projects using [Tape](https://
...
+
 If we were unit testing the route handler, our tests would stub the logger, and ignore the interactions with it, testing only whether or not the route responded appropriately to the faked request.
 
+如果我们队路由处理器做但也测试。我们不会管记录器，或者与它有任何交互。我们只会对路由做假的请求来测试。
```
Comment (marcmoore) [inline TODO/javascript-testing-unit-functional-integration.md]: 这句问题比较多，译者再看看吧。

File: TODO/javascript-testing-unit-functional-integration.md [hunk 14]
```diff
@@ -150,23 +231,42 @@ Here are some simple unit test examples from real projects using [Tape](https://
...
+
 Let’s look at this in more depth. The route handler is a factory function which uses dependency injection to inject the logger into the route handler. Let’s look at the signature (See the [rtype docs](https://github.com/ericelliott/rtype) for help reading signatures):
 
+进一步说。路由处理器是一个工厂方法，靠依赖注入来注入进记录器里。让我们来看看 signature：
```
Comment (marcmoore) [inline TODO/javascript-testing-unit-functional-integration.md]: route handler 作为一种工厂函数通过依赖注入将logger注入进去。我们来看一段批注

File: TODO/javascript-testing-unit-functional-integration.md [hunk 15]
```diff
@@ -192,38 +292,66 @@ Let’s see how we can test this:
...
 
+## 功能测试
+
 Functional tests are automated tests which ensure that your application does what it’s supposed to do from the point of view of the user. Functional tests feed input to the user interface, and make assertions about the output that ensure that the software responds the way it should.
 
+功能测试是确保您的应用程序从用户的角度来看正常运行的自动化测试。功能测试测试用户的界面，输入和输出，确保软件按照期望方式做出响应。
+
 Functional tests are sometimes called end-to-end tests because they test the entire application, and it’s hardware and networking infrastructure, from the front end UI to the back end database systems. In that sense, functional tests are also a form of integration testing, ensuring that machines and component collaborations are working as expected.
 
+功能测试有时被称为端到端测试，因为他们测试整个应用程序，已经它相关的硬件和网络基础设施，从前端 UI 到后端数据库系统。在这个意义上，功能测试也是一种集成测试，确保机器和组件的都按期望工作。
```
Comment (marcmoore) [inline TODO/javascript-testing-unit-functional-integration.md]: 译者请读一遍，有些错别字。

File: TODO/javascript-testing-unit-functional-integration.md [hunk 16]
```diff
@@ -240,44 +368,83 @@ My favorite is [Nightwatch.js](http://nightwatchjs.org/). Here’s what a simple
...
 
+维护一份自动化的功能测试 - 比方说烟雾测试，是很重要的。测试你应用中所有的重要功能。那些用户在日常操作中会遇到的请求。
+
 Smoke tests are not the only use for functional tests, but in my opinion, they’re the most valuable.
 
+烟雾测试不是功能测试的唯一作用， 但是在我看来，却是最有意义的
+
 ## What Is Continuous Delivery?
 
+## 为什么要持续迭代发布产品？
```
Comment (marcmoore) [inline TODO/javascript-testing-unit-functional-integration.md]: 什么是持续部署

File: TODO/javascript-testing-unit-functional-integration.md [hunk 17]
```diff
@@ -240,44 +368,83 @@ My favorite is [Nightwatch.js](http://nightwatchjs.org/). Here’s what a simple
...
+5. 部署
+6. 维护
 
 It’s called waterfall because if you chart it with time running from right to left, it looks like a waterfall cascading from one task to the next. In other words, in theory, you can’t really do these things concurrently.
 
+它之所以被称为瀑布，是因为如果你记录它从右到左的运行的时间，它看起来像从一个任务到下一个级联的瀑布。换句话说，在理论上你不能同时做这些事情。
+
 In theory. In reality, a lot of project scope is discovered as the project is being developed, and scope creep often leads to disastrous project delays and rework. Inevitably, the business team will also want “simple changes” made after delivery without going through the whole expensive, time-consuming waterfall process again, which frequently results in an endless cycle of change management meetings and production hot fixes.
 
+理论上（译注：这是作者打错了？怎么又是理论上又是实际上...），实际上，很多正在开发的项目的需求是在开发中才被发现的，而范围的变化常常导致灾难性的工程延误和返工。不可避免地，业务团队也会想"简单的改变”在发布后的产品，而不打算通过整个昂贵、 耗时的瀑布式的过程，这经常导致在无限循环的变化管理会议和产品热修复。
```
Comment (marcmoore) [inline TODO/javascript-testing-unit-functional-integration.md]: scope creep 翻译成“需求变更”

File: TODO/javascript-testing-unit-functional-integration.md [hunk 18]
```diff
@@ -240,44 +368,83 @@ My favorite is [Nightwatch.js](http://nightwatchjs.org/). Here’s what a simple
...
 
+一个理论上的瀑布过程可能只是一个神话。在我的长长的职业生涯中，我与数百家企业，进行了磋商，但我从没见在真正的生产中见过完美的瀑布。典型的瀑布的发布周期可能会是几个月或几年。
+
 ## The Continuous Delivery Solution
 
+## 持续迭代的解决办法
+
 Continuous delivery is a development methodology that acknowledges that scope is uncovered as the project progresses, and encourages incremental improvements to software in short cycles that ensure that software can be released at any time without causing problems.
 
+持续迭代法不是一种开发方法，承认需求是随着项目的进展而被挖掘的，鼓励在短周期内增量改进软件，并确保不会导致问题，在任何时候软件发布。
```
Comment (marcmoore) [inline TODO/javascript-testing-unit-functional-integration.md]: 这一段翻译的有点奇怪了，译者请再读一下。

File: TODO/javascript-testing-unit-functional-integration.md [hunk 19]
```diff
@@ -240,44 +368,83 @@ My favorite is [Nightwatch.js](http://nightwatchjs.org/). Here’s what a simple
...
+
 In contrast to the waterfall method, I’ve seen the continuous delivery process running smoothly at dozens of organizations — but I’ve never seen it work anywhere without a quality array of test suites that includes both unit tests and functional tests, and frequently includes integration tests, as well.
 
+对比瀑布方案，我在无数的企业组织中都见过迭代开发顺利进行 —— 但我从没见过哪一个是在没有单元和功能的测试组件的情况下完成的，通常，测试组件也会包括集成测试。
+
 Hopefully now you have everything you need to get started on your continuous delivery foundations.
 
+希望我的这篇文章告诉了你开始迭代发布所需要知道的所有内容。
+
 ## Conclusion
```
Comment (marcmoore) [inline TODO/javascript-testing-unit-functional-integration.md]: 总结

**Comment by marcmoore** [comment]
@sqrthree 校对完了。Phew~~~

**Comment by sqrthree** [comment]
@marcmoore 好的，辛苦啦

**Comment by Tina92** [comment]
@sqrthree 校对申领

**Comment by sqrthree** [comment]
好的👌🏻

**Comment by sqrthree** [comment]
@Tina92 别忘了来校对哈

File: TODO/javascript-testing-unit-functional-integration.md [hunk 20]
```diff
@@ -1,116 +1,197 @@
...
+# JavaScript 测试︰ 单元 vs 功能 vs 集成测试
 
 Unit tests, integration tests, and functional tests are all types of automated tests which form essential cornerstones of continuous delivery, a development methodology that allows you to safely ship changes to production in days or hours rather than months or years.
 
+自动测试包括了单元测试、 集成测试和功能测试。自动测试是一种开发方法，它是持续不断迭代发布产品的基石，使你生产环境里的发布的能够在数小时，数天之内就发生改变，而不是按月或是按年而计。
+
 Automated tests enhance software stability by catching more errors before software reaches the end user. They provide a safety net that allows developers to make changes without fear that they will unknowingly break something in the process.
 
-## The Cost of Neglecting Tests
+自动测试通过捕捉更多的错误，增强了软件到达用户之前的稳定性。就好比是一张防护网一样，使开发者们在做更改的时候不必担心弄坏了东西。
```
Comment (Tina92) [inline TODO/javascript-testing-unit-functional-integration.md]: - 使开发者们在做更改的时候不必担心弄坏了东西。

- 使开发者们在做更改的时候不必担心引发未知的错误。

File: TODO/javascript-testing-unit-functional-integration.md (hunk 2 above)
Comment (Tina92) [inline TODO/javascript-testing-unit-functional-integration.md]: 确实与本文无关

File: TODO/javascript-testing-unit-functional-integration.md [hunk 21]
```diff
@@ -1,116 +1,197 @@
...
 *   [The Future of Programming: WebAssembly & Life After JavaScript](https://www.sitepoint.com/future-programming-webassembly-life-after-javascript/?utm_source=sitepoint&utm_medium=relatedinline&utm_term=&utm_campaign=relatedauthor)
-
-
-
-
+（上面这段文字和本文内容无关，故我打算删掉，请校对过目。）
 
 Contrary to popular intuition, maintaining a quality test suite can dramatically enhance developer productivity by catching errors immediately. Without them, end users encounter more bugs, which can lead to increased reliance on customer service, quality assurance teams, and bug reports.
 
+与直觉相反，维护一个高质量的测试套件能够极大地提高开发人员的效率，（因为它使得开发人员）能够立即发现开发中的错误。反过来说，如果没有这些套件，用户会遇到更多的 Bug，从而导致需要投入更多的在用户服务，质量保证和错误报告上面。
```
Comment (Tina92) [inline TODO/javascript-testing-unit-functional-integration.md]: end users翻译成终端用户更好一些吧～～

File: TODO/javascript-testing-unit-functional-integration.md [hunk 22]
```diff
@@ -1,116 +1,197 @@
...
+
 Test Driven Development takes a little more time up front, but bugs that reach customers cost more in many ways:
+测试驱动开发（TDD）在前期需要更多的时间，但是一旦 Bug 到了用户那里，你会付出更多的代价：
```
Comment (Tina92) [inline TODO/javascript-testing-unit-functional-integration.md]: - 需要更多的时间

- 需要稍微多一点的时间

File: TODO/javascript-testing-unit-functional-integration.md [hunk 23]
```diff
@@ -1,116 +1,197 @@
...
+所有的错误报告都必须被 QA 或者开发者亲自验证。
+修补这些 Bug 会耗费你大量的时间，因为它导致你必须停下手头的工作。粗略估计每一个 Bug 都将浪费你 20分钟的时间，而这还没有算你修补它们的时间。
+有时候，诊断这些 Bug 的人并不是开发它们的人，这导致了开发人员对代码的不熟悉。
+机会成本：开发团队必须等到 Bug 被修补以后，才能继续按照计划开发。
+
 The cost of a bug that makes it into production is many times larger than the cost of a bug caught by an automated test suite. In other words, TDD has an overwhelmingly positive ROI.
+在生产环境里的 Bug 使你付出的代价往往要数倍于在自动化测试时发现的。换句话说，在计算算投资与回报时，测试驱动开发（TDD）具有压倒性的优势。
 
 ## Different Types of Tests
+## 不同的测试类型
```
Comment (Tina92) [inline TODO/javascript-testing-unit-functional-integration.md]: 不同类型的测试

File: TODO/javascript-testing-unit-functional-integration.md [hunk 24]
```diff
@@ -1,116 +1,197 @@
...
 
+我观察到的其中一个问题就是，现有套件混淆了单元，功能和集成测试。他们之间没有明显的区别。
+
 The result was a test suite that was not particularly well suited for anything.
 
+其结果是现有测试套件不适合任何一个测试套件。
+
 ## Roles Tests Play in Continuous Delivery
 
+## 在持续迭代发布中它们所扮演的角色
```
Comment (Tina92) [inline TODO/javascript-testing-unit-functional-integration.md]: 在持续迭代发布中它们的角色

File: TODO/javascript-testing-unit-functional-integration.md [hunk 25]
```diff
@@ -1,116 +1,197 @@
...
+
 Each type of test has a unique role to play. You don’t choose between unit tests, functional tests, and integration tests. Use all of them, and make sure you can run each type of test suite in isolation from the others.
 
+每种类型的测试可以发挥其独特的作用。你不用选择在单元测试、 功能测试和集成测试中做选择。因为你会使用它们所有，并确保您也可以独立的运行每种类型的测试套件。
```
Comment (Tina92) [inline TODO/javascript-testing-unit-functional-integration.md]: 个人觉得这边译者翻译得更好

File: TODO/javascript-testing-unit-functional-integration.md (hunk 5 above)
Comment (Tina92) [inline TODO/javascript-testing-unit-functional-integration.md]: 个人觉得可能翻译为：边缘情况

File: TODO/javascript-testing-unit-functional-integration.md (hunk 6 above)
Comment (Tina92) [inline TODO/javascript-testing-unit-functional-integration.md]: 针对终端用户

File: TODO/javascript-testing-unit-functional-integration.md (hunk 8 above)
Comment (Tina92) [inline TODO/javascript-testing-unit-functional-integration.md]: 不要用网络用语吧～

Comment (Tina92) [inline TODO/javascript-testing-unit-functional-integration.md]: 什么是

File: TODO/javascript-testing-unit-functional-integration.md [hunk 26]
```diff
@@ -1,116 +1,197 @@
...
 
+1. 什么组件被测试 ing
+2. 你期望什么行为
+3. 实际是什么结果
+4. 你期望什么结果
+5. 为什么会这样
+
 The first four questions should be visible in the failure report. The last question should be clear from the test’s implementation. Some assertion types are not capable of answering all those questions in a failure report, but most `equal`, `same`, or `deepEqual` assertions should. In fact, if those were the only assertions in any assertion library, most test suites would probably be better off. Simplify.
 
+前四个问题应在故障报告中清晰可见，而最后的那个问题应该从测试的执行很清楚的中找到。当然，有一些断言的类型是不能回答所有这些问题的，但大部分的问题 ‘equal’、 ’same'，或者 'deepEqual’ 都应改可以做到。事实上，如果那些断言是现有断言库里的唯一断言，现存的大多数测试套件可能会更好。大道至简！
```
Comment (Tina92) [inline TODO/javascript-testing-unit-functional-integration.md]: - 当然，有一些断言的类型是不能回答所有这些问题的

- 当然，在一份不合格的报告中有一些assertion的类型是不能回答所有问题的

File: TODO/javascript-testing-unit-functional-integration.md [hunk 27]
```diff
@@ -150,23 +231,42 @@ Here are some simple unit test examples from real projects using [Tape](https://
 
 ## Integration Tests
 
+## 集成测试
+
 Integration tests ensure that various units work together correctly. For example, a Node route handler might take a logger as a dependency. An integration test might hit that route and test that the connection was properly logged.
 
+集成测试确保各组件一起正常工作。例如，节点路由处理程序可能需要一个记录器（logger）作为依赖。集成测试可测试路由和连接被正确的记录。
```
Comment (Tina92) [inline TODO/javascript-testing-unit-functional-integration.md]: - 节点路由处理程序可能需要一个记录器（logger）作为依赖

- 节点路由处理程序可能需要依赖记录器（logger）。

File: TODO/javascript-testing-unit-functional-integration.md [hunk 28]
```diff
@@ -240,44 +368,83 @@ My favorite is [Nightwatch.js](http://nightwatchjs.org/). Here’s what a simple
...
 
+烟雾测试不是功能测试的唯一作用， 但是在我看来，却是最有意义的
+
 ## What Is Continuous Delivery?
 
+## 为什么要持续迭代发布产品？
+
 Prior to the continuous delivery revolution, software was released using a waterfall process. Software would go through the following steps, one at a time. Each step had to be completed before moving on to the next:
 
+在持续交付革命之前，软件发布都是使用瀑布过程。软件发布通过以下步骤，一次一个，每一步不得移到下一步之前完成︰
```
Comment (Tina92) [inline TODO/javascript-testing-unit-functional-integration.md]: - 每一步不得移到下一步之前完成︰

- 每一步必须在移到下一步之前完成︰

File: TODO/javascript-testing-unit-functional-integration.md [hunk 29]
```diff
@@ -1,116 +1,197 @@
...
+**开发阶段**，主要是程序员反馈。这时单元测试很有用。
+**在中间阶段**，主要是能够在发现问题时立刻停下来。这时各种测试都很有用。
+**在生产环境**，主要是运行测试套件中一个叫做「冒烟测试」的子集，确保部署的时候没有弄坏什么东西。
 
 Which Test Types Should You Use? All of Them.
+如果你问我该使用那个测试？**所有的。**
 
 In order to understand how different tests fit in your software development process, you need to understand that each kind of test has a job to do, and those tests roughly fall into three broad categories:
 
+为了了解如何在您的软件开发过程选择不同测试，你需要了解每种测试它的工作要做，那些测试大致可分为三大类︰
```
Comment (wild-flame) [inline TODO/javascript-testing-unit-functional-integration.md]: 这句话貌似是打字的时候打漏字了的……后面的时间有点紧，我翻译完了都没有仔细读就 pull request 了。

我貌似是想说「你需要了解每种测试都有它的工作要做」

File: TODO/javascript-testing-unit-functional-integration.md [hunk 30]
```diff
@@ -240,44 +368,83 @@ My favorite is [Nightwatch.js](http://nightwatchjs.org/). Here’s what a simple
...
 Hopefully now you have everything you need to get started on your continuous delivery foundations.
 
+希望我的这篇文章告诉了你开始迭代发布所需要知道的所有内容。
+
 ## Conclusion
 
 As you can see, each type of test has an important part to play. Unit tests for fast developer feedback, integration tests to cover all the corner cases of component integrations, and functional tests to make sure everything works right for the end users.
 
-How do you use automated tests in your code, and how does it impact your confidence and productivity? Let me know in the comments.
+正如你所看到的，每种类型的测试发挥了重要作用。单元测试能够快速的反馈开发者，集成测试会覆盖所有的角落的组件，而功能测试确保一切最终用户的那里的情况一切正常。
```
Comment (Tina92) [inline TODO/javascript-testing-unit-functional-integration.md]: - 功能测试确保一切最终用户的那里的情况一切正常。

- 功能测试确保一切在终端用户的那里运行正常。

**Comment by Tina92** [comment]
@sqrthree  校对完成

File: TODO/javascript-testing-unit-functional-integration.md (hunk 7 above)
Comment (wild-flame) [inline TODO/javascript-testing-unit-functional-integration.md]: 「为了实现这个目的，单元测试必须很快，也就是说，在单元测试里，一切异步的操作都应该避免。」这样应该好些

**Comment by wild-flame** [comment]
好的 @Tina92 @marcmoore 谢啦~~

**Comment by wild-flame** [comment]
对了。我一直很纠结 The Continuous Delivery  应该翻译成 「持续迭代」还是「持续发布」。或者应该是「持续迭代发布」？我觉得三者的意思应该是一样的，但是不知道怎么译才好。

p.s.又或者「迭代发布」？

File: TODO/javascript-testing-unit-functional-integration.md [hunk 31]
```diff
@@ -1,116 +1,197 @@
...
+
 The cost of a bug that makes it into production is many times larger than the cost of a bug caught by an automated test suite. In other words, TDD has an overwhelmingly positive ROI.
+在生产环境里的 Bug 使你付出的代价往往要数倍于在自动化测试时发现的。换句话说，在计算算投资与回报时，测试驱动开发（TDD）具有压倒性的优势。
 
 ## Different Types of Tests
+## 不同的测试类型
 
 The first thing you need to understand about different types of tests is that they all have a job to do. They play important roles in continuous delivery.
 
+你需要了解的第一件事情就是，每一种测试都有它自己的作用。他们都在软件的持续迭代发布中起了重要的作用。
```
Comment (marcmoore) [inline TODO/javascript-testing-unit-functional-integration.md]: 和72行应该用相同的翻译方式吧。Each type of test has a unique role to play.

**Comment by marcmoore** [comment]
@wild-flame 持续发布。迭代指的是iteration。关注的是一个过程。Delivery是一个动作。

**Comment by sqrthree** [comment]
@wild-flame 这篇文章调整的如何了？

File: TODO/javascript-testing-unit-functional-integration.md (hunk 5 above)
Comment (wild-flame) [inline TODO/javascript-testing-unit-functional-integration.md]: 我觉得可能是边缘情况~~

File: TODO/javascript-testing-unit-functional-integration.md (hunk 29 above)
Comment (wild-flame) [inline TODO/javascript-testing-unit-functional-integration.md]: 这句话应该也可以译成「所扮演的角色」

**Comment by wild-flame** [comment]
@sqrthree 好了。可以 merge了

File: TODO/javascript-testing-unit-functional-integration.md [hunk 32]
```diff
@@ -1,115 +1,103 @@
...
-The first thing you need to understand about different types of tests is that they all have a job to do. They play important roles in continuous delivery.
+如果你问我该使用那个测试？**所有的。**
 
-A while back, I was consulting on an ambitious project where the team was having a hard time building a reliable test suite. Because it was hard to use and understand, it rarely got used or maintained.
+为了了解如何在您的软件开发过程选择不同测试，你需要了解每种测试所扮演的角色，那些测试大致可分为三大类︰
 
-One of the problems I observed with the existing test suite is that it confused unit tests, functional tests, and integration tests. It made absolutely no distinction between any of them.
+*   用户体验测试 （针对最终用户）
+*   开发 API 测试 （针对开发人员）
+*   基础设施测试 (负载测试、 网络集成测试等......)
```
Comment (sqrthree) [inline TODO/javascript-testing-unit-functional-integration.md]: image

Comment (sqrthree) [inline TODO/javascript-testing-unit-functional-integration.md]: 没有对齐哈

File: TODO/javascript-testing-unit-functional-integration.md [hunk 33]
```diff
@@ -1,115 +1,103 @@
...
-4.  What is the expected result?
-5.  How is the behavior reproduced?
+1. 什么组件正在被测试
+2. 你期望什么行为
+3. 实际是什么结果
+4. 你期望什么结果
+5. 如何重现
 
-The first four questions should be visible in the failure report. The last question should be clear from the test’s implementation. Some assertion types are not capable of answering all those questions in a failure report, but most `equal`, `same`, or `deepEqual` assertions should. In fact, if those were the only assertions in any assertion library, most test suites would probably be better off. Simplify.
+前四个问题应在故障报告中清晰可见，而最后的那个问题应该从测试的执行很清楚的中找到。当然，在一份不合格的报告中有一些assertion的类型是不能回答所有问题的,但大部分的问题 ‘equal’、 ’same'，或者 'deepEqual’ 都应该可以做到。事实上，如果那些断言是现有断言库里的唯一断言，现存的大多数测试套件可能会更好。大道至简！
```
Comment (sqrthree) [inline TODO/javascript-testing-unit-functional-integration.md]: 『assertion』英文前后的空格

File: TODO/javascript-testing-unit-functional-integration.md [hunk 34]
```diff
@@ -148,24 +136,24 @@ Here are some simple unit test examples from real projects using [Tape](https://
...
-In this case, we have two units under test:
+这里有两个组件同时被测试了：
 
-1.  The route handler
-2.  The logger
+1. 路由处理器（Route handler）
+2. 记录器（Logger）
 
-If we were unit testing the logger, our tests wouldn’t invoke the route handler, or know anything about it.
+如果我们对 logger 做单元测试，这些测试是不会调用到route handler，或者说根本就不知道还有个handler。
```
Comment (sqrthree) [inline TODO/javascript-testing-unit-functional-integration.md]: 空格

**Comment by wild-flame** [comment]
@sqrthree 好了

**Comment by sqrthree** [comment]
@wild-flame 已经 merge 啦~ 快快麻溜分享到掘金然后给我发下链接，方便及时添加积分哟
